### ISSUE TITLE
DCM matrix builter

### ISSUE THREAD

**Issue by Marchesini010126** [tudat-team/tudat-space#65]
Hi

I have written two simple functions that could be useful in the future if you want to add them. I hope they can be of any help and let me know if you want me to modify them 

```
import numpy as np

def split_history(state_history:dict,n_bodies:int):
    
    """
    Parameters
    -----------
    
    state_history        (dict ) :  dictionary containing the state_history
                                    of N propagated bodies.
                                    The dictionary is defined as a standard 
                                    state_history following the TUdat convention. 
                                    
                                    the keys of the dictionary are the time stamps 
                                    corresponding to each state and the argument is 
                                    the state of the body at the given time stamp
                                    
                                    
    n_bodies            (int)     : number of propagated bodies
    
    Returns
    -----------
    
    state_hostory_book (list)   : list of state_history dictionaries for each propagated body.
                                  The order of the list is given directly by the order 
                                  in which the bodies are propagated.
    
    Description
    -----------
    Creates a list of state histories based on the unified state_history
    from the propagation of n_bodies 
    
    """
    
    state_history_book = []
    time               = [key for key in state_history.keys() ]
    n_variables        = len(state_history[time[0]])
    step               = int(n_variables/n_bodies)
    
    for n in range(n_bodies) :
        current_history = dict()
        
        for t in time:
           current_history[t]= state_history[t][step*n:step*n+step]
        
        state_history_book.append(current_history)
    
    return state_history_book

def vector2matrix(flat_matrix:np.ndarray) :
    
    """
    Parameters
    -----------
    
    flat_matrix     (np.ndarray ) : vector containing a flatttened rotation matrix 
                                    following the TuDat standards
                                    a rotation matrix is returned as a nine-entry 
                                    vector in the dependent variable output, where entry
                                    (i,j) of the matrix is stored in entry (3i+j)  of the vector
                                    with i,j = 0,1,2
                                    
    
                                    
    
    Returns
    -----------
    
    rotation_matrix      (np.ndarray)     : returns (3x3) Rotation matrix 
                                            (orthogonal matrix)
                                    
                                   
    
    Descriptiom
    -----------
    
    Retruns rotation matrix from its flattened vectorial version.
    Check for details :
    https://tudatpy.readthedocs.io/en/latest/dependent_variable.html#tudatpy.numerical_simulation.propagation_setup.dependent_variable.inertial_to_body_fixed_rotation_frame
    
    """
    
    rotations_matrix = np.empty((3,3))
    
    for i in range(3) :
        for j in range(3) :
           rotations_matrix[i,j]=flat_matrix[3*i+j]
   
    return rotations_matrix
```

**Comment by DominicDirkx** [comment]
Hi Greg, this looks useful :) Thanks!

@Marchesini010126 If we were to add this now, we would copy paste this into he code and push it. If you formulate it as a pull request, your name would also show up in the tudat git commit history. In case you would want that, for us it makes little difference (I think)

@gaffarelj Could you add these to a future tudatpy update?

**Comment by gaffarelj** [comment]
Hi all :)

I can take care of adding these functions to tudatpy, yes. Looking through them, I will then put them inside the tudatpy.util module.

@Marchesini010126 tell me if you want to take care of it trough a push request, otherwise I will make it myself, and mention your username in the commit message.

**Comment by gaffarelj** [comment]
Hi @Marchesini010126!

Small update: I think that there is an incorrect assumption with your `split_history` function. It looks like you assume your state to be _[translational body A, mass body A, translational body B, mass body B]_, where it is in fact _[translational body A, translational body B, mass body A, mass body B]_ (as an example).
So your function, as it is, would work fine if you propagate only translational dynamics, but not if you propagate any other state type (like mass, rotational, etc). 

I am making the appropriate changes to your function (and tudatpy) so that this functionality can still be implemented, but that will take just a little bit longer (though it's well underway already).

**Comment by Marchesini010126** [comment]
Hey 

Yes you are right
I mainly work with translational dynamics and this is the reason why I had made this assumption.

Really grateful I watched this issue and let me know if you need anything!
Great catch 

Have a nice day
Gregorio Marchesini

Personal Email
***@***.***

Academic Email
***@***.***

**Comment by gaffarelj** [comment]
I implemented both of these functions and made [this pull request](https://github.com/tudat-team/tudatpy/pull/66) accordingly. @DominicDirkx can you review it?

Thanks again for these additions @Marchesini010126 :)

**Comment by gaffarelj** [comment]
These new features have been merged with `tudatpy/develop` with commit [c56ab7f](https://github.com/tudat-team/tudatpy/commit/c56ab7fdaf4ff6fa3ce82004abe67e9c4b0fb3d3).
They will be included in the next conda release of tudatpy, and in the online documentation as soon as possible (there is a little fix needed for the docs to be updatable again).

**Comment by gaffarelj** [comment]
Just to confirm, the new API docs (that are now building again) contain the two new features mentioned in this issue:

- [util.split_history](https://py.api.tudat.space/en/latest/util.html#tudatpy.util.split_history)
- [util.vector2matrix](https://py.api.tudat.space/en/latest/util.html#tudatpy.util.vector2matrix)
